pkg/terraform/errors/errors.go

```diff
@@ -16,40 +16,156 @@ limitations under the License.
 
 package errors
 
+import (
+	"fmt"
+	"strings"
+
+	jsoniter "github.com/json-iterator/go"
+	"github.com/pkg/errors"
+)
+
+const (
+	levelError = "error"
+)
+
+type tfError struct {
+	logs    []byte
+	tfLogs  []*TerraformLog
+	cause   error
+	message string
+}
+
 type applyFailed struct {
-	log string
+	*tfError
+}
+
+// TerraformLog represents relevant fields of a Terraform CLI JSON-formatted log line
+type TerraformLog struct {
+	Level      string        `json:"@level"`
+	Message    string        `json:"@message"`
+	Diagnostic LogDiagnostic `json:"diagnostic"`
+}
+
+// LogDiagnostic represents relevant fields of a Terraform CLI JSON-formatted
+// log line diagnostic info
+type LogDiagnostic struct {
+	Severity string `json:"severity"`
+	Summary  string `json:"summary"`
+	Detail   string `json:"detail"`
+}
+
+func (t *tfError) Error() string {
+	if t == nil {
+		return ""
+	}
+
+	messages := make([]string, 0, len(t.tfLogs))
+	for _, l := range t.tfLogs {
+		// only use error logs
+		if l == nil || l.Level != levelError {
+			continue
+		}
+		m := l.Message
+		if l.Diagnostic.Severity == levelError && l.Diagnostic.Summary != "" {
+			m = fmt.Sprintf("%s: %s", l.Diagnostic.Summary, l.Diagnostic.Detail)
+		}
+		messages = append(messages, m)
+	}
+	return fmt.Sprintf("%s: %s", t.message, strings.Join(messages, "\n"))
+}
+
+func (t *tfError) Unwrap() error {
+	if t == nil {
+		return nil
+	}
+	return t.cause
 }
 
-func (a *applyFailed) Error() string {
-	return a.log
+func newTFError(message string, cause error, logs []byte) (string, *tfError) {
+	tfError := &tfError{
+		logs:    logs,
+		cause:   cause,
+		message: message,
+	}
+
+	tfLogs, err := parseTerraformLogs(logs)
+	if err != nil {
+		return err.Error(), tfError
+	}
+	tfError.tfLogs = tfLogs
+	return "", tfError
 }
 
-// NewApplyFailed returns a new apply failure error with given logs.
-func NewApplyFailed(log string) error {
-	return &applyFailed{log: log}
+func parseTerraformLogs(logs []byte) ([]*TerraformLog, error) {
+	logLines := strings.Split(string(logs), "\n")
+	tfLogs := make([]*TerraformLog, 0, len(logLines))
+	for _, l := range logLines {
+		log := &TerraformLog{}
+		l := strings.TrimSpace(l)
+		if l == "" {
+			continue
+		}
+		if err := jsoniter.ConfigCompatibleWithStandardLibrary.UnmarshalFromString(l, log); err != nil {
+			return nil, err
+		}
+		tfLogs = append(tfLogs, log)
+	}
+	return tfLogs, nil
+}
+
+// WrapTFError returns a new Terraform CLI failure error with given logs.
+func WrapTFError(message string, cause error, logs []byte) error {
```

Comment:
Seems like we can have `WrapPlanFailed` and `WrapRefreshFailed` instead of this generic one since command types we run are determined and is a very limited set.

Reply:
Yes, correct. I did not see any current use case for specialized errors for them but it should be straightforward.